crates/bevy_render/src/shader/shader_defs.rs

```diff
@@ -60,55 +52,45 @@ impl ShaderDef for Option<Handle<Texture>> {
 }
 
 /// Updates [RenderPipelines] with the latest [ShaderDefs]
-pub fn shader_defs_system<T>(mut query: Query<(&T, &mut RenderPipelines)>)
+pub fn shader_defs_system<T>(mut query: Query<(&T, &mut RenderPipelines), Changed<T>>)
 where
     T: ShaderDefs + Send + Sync + 'static,
 {
-    for (shader_defs, mut render_pipelines) in query.iter_mut() {
-        for shader_def in shader_defs.iter_shader_defs() {
-            for render_pipeline in render_pipelines.pipelines.iter_mut() {
-                render_pipeline
-                    .specialization
-                    .shader_specialization
-                    .shader_defs
-                    .insert(shader_def.to_string());
-            }
-        }
-    }
+    query.iter_mut().for_each(update_render_pipelines)
 }
 
-/// Clears each [RenderPipelines]' shader defs collection
-pub fn clear_shader_defs_system(mut query: Query<&mut RenderPipelines>) {
-    for mut render_pipelines in query.iter_mut() {
+/// Insert defined shader defs and remove undefined ones from render pipelines.
+fn update_render_pipelines<T>(q: (&T, Mut<RenderPipelines>))
+where
+    T: ShaderDefs + Send + Sync + 'static,
+{
+    let (shader_defs, mut render_pipelines) = q;
+    for (shader_def, is_defined) in shader_defs.iter_shader_defs() {
         for render_pipeline in render_pipelines.pipelines.iter_mut() {
-            render_pipeline
+            let shader_defs = &mut render_pipeline
                 .specialization
                 .shader_specialization
-                .shader_defs
-                .clear();
+                .shader_defs;
+            let s = shader_def.to_string();
+            if is_defined {
+                shader_defs.insert(s);
+            } else {
+                shader_defs.remove(&s);
+            }
         }
     }
 }
 
 /// Updates [RenderPipelines] with the latest [ShaderDefs] from a given asset type
 pub fn asset_shader_defs_system<T: Asset>(
     assets: Res<Assets<T>>,
-    mut query: Query<(&Handle<T>, &mut RenderPipelines)>,
+    mut query: Query<(&Handle<T>, &mut RenderPipelines), Changed<Handle<T>>>,
 ) where
     T: ShaderDefs + Send + Sync + 'static,
 {
-    for (asset_handle, mut render_pipelines) in query.iter_mut() {
-        if let Some(asset_handle) = assets.get(asset_handle) {
-            let shader_defs = asset_handle;
-            for shader_def in shader_defs.iter_shader_defs() {
-                for render_pipeline in render_pipelines.pipelines.iter_mut() {
-                    render_pipeline
-                        .specialization
-                        .shader_specialization
-                        .shader_defs
-                        .insert(shader_def.to_string());
-                }
-            }
-        }
-    }
+    query
+        .iter_mut()
+        // (Handle<T>, _) -> (&T, _)
+        .filter_map(|(h, p)| assets.get(h).map(|a| (a, p)))
```

Comment:
I expect this to behave incorrectly when an asset isn't loaded yet. Ex:

1. begin loading a "shader def" asset and assign its handle to a component 
2. this system runs. the entity with the handle shows up in the query iteration because the handle was changed
3. it gets filtered out because the asset isn't loaded yet
4.  later, when the asset finishes loading, this system doesn't run on the entity we skipped, because the Changed flag is no longer set.